vllm/worker/worker.py

```diff
@@ -195,62 +195,73 @@ def determine_num_available_blocks(self) -> Tuple[int, int]:
-        if self.model_runner.lora_manager:
-            self.model_runner.remove_all_loras()
-        gc.collect()
-
-        return num_gpu_blocks, num_cpu_blocks
+            # Check for any memory left around that may have been allocated
+            # on thegpu outside of `torch`. NCCL operations, for example,
+            # can use a few GB during a forward pass
+            torch.cuda.empty_cache()
+            torch_allocated_bytes = torch.cuda.memory_stats(
+            )["allocated_bytes.all.current"]
+            total_allocated_bytes = torch.cuda.mem_get_info(
+            )[1] - torch.cuda.mem_get_info()[0]
+
+            non_torch_allocations = (total_allocated_bytes -
+                                     torch_allocated_bytes)
+            if non_torch_allocations > 0:
+                peak_memory += non_torch_allocations
+
+            available_kv_cache_memory = (
+                total_gpu_memory * self.cache_config.gpu_memory_utilization -
+                peak_memory)
+
+            # Calculate the number of blocks that can be allocated with the
+            # profiled peak memory.
+            cache_block_size = self.get_cache_block_size_bytes()
+            if cache_block_size == 0:
+                num_gpu_blocks = 0
+                num_cpu_blocks = 0
+            else:
+                num_gpu_blocks = int(available_kv_cache_memory //
+                                     cache_block_size)
+                num_cpu_blocks = int(self.cache_config.swap_space_bytes //
+                                     cache_block_size)
+            num_gpu_blocks = max(num_gpu_blocks, 0)
+            num_cpu_blocks = max(num_cpu_blocks, 0)
+
+            logger.info(
+                "Memory profiling results: total_gpu_memory=%.2fGiB"
+                " initial_memory_usage=%.2fGiB peak_torch_memory=%.2fGiB"
+                " memory_usage_post_profile=%.2fGib"
+                " non_torch_memory=%.2fGiB kv_cache_size=%.2fGiB"
+                " gpu_memory_utilization=%.2f", total_gpu_memory / (1024**3),
+                (total_gpu_memory - free_memory_pre_profile) / (1024**3),
+                (peak_memory - non_torch_allocations) / (1024**3),
+                total_allocated_bytes / (1024**3),
+                non_torch_allocations / (1024**3),
+                available_kv_cache_memory / (1024**3),
+                self.cache_config.gpu_memory_utilization)
+
+            # Final cleanup
+            if self.model_runner.lora_manager:
+                self.model_runner.remove_all_loras()
+            gc.collect()
+
+            return num_gpu_blocks, num_cpu_blocks
+        except Exception:
+            # Provide fallback values based on total GPU memory with a
+            # 70% utilization safety margin
+            cache_block_size = self.get_cache_block_size_bytes()
+            safe_gpu_blocks = int((total_gpu_memory * 0.7) //
+                                  cache_block_size) if cache_block_size else 0
```

Comment:
1. Better to have proper a message to avoid overriding user configuration silently.
2. I'm not sure if heuristic fallback is a desired behavior, because 0.7 may not be an optimal value and may result in sub-optimal performance (or OOM if it's still too large?) with confusion. For example, if users want to tune the GPU memory utilization, they may see the following trend:

Util | #Blocks | Note
--|--|--
0.9 | 1000 | OOM and fallback to 0.7
0.8 | 1200 | The optimal value in this case
0.7 | 1000 | Just about it
0.6 | 900 | Safe